### ISSUE TITLE
Rewrite-maven: ChangeParentPom recipe does not work correctly with multi-module maven projects

### ISSUE THREAD

**Issue by ferblaca** [openrewrite/rewrite#1753]
I have problems applying the `ChangeParentPom` recipe in multi-module maven projects where the goal is to modify the parent version of the root module of type `pom`.

It seems that the problem is related to the number of modules the maven project contains...

For example, I have created a Recipe in the following [project ](https://github.com/ferblaca/OpenRewriteRecipe) with the goal of modifying the Spring Boot parent to version 2.6.7:

```
---
type: specs.openrewrite.org/v1beta/recipe
name: com.yourorg.ChangeParentPomExample
displayName: Change Maven Parent Pom example
recipeList:
  - org.openrewrite.maven.ChangeParentPom:
      oldGroupId: org.springframework.boot
      newGroupId: org.springframework.boot
      oldArtifactId: spring-boot-starter-parent
      newArtifactId: spring-boot-starter-parent
      newVersion: 2.6.7
      versionPattern:
      allowVersionDowngrades: false
```

And on the other hand, I have a [demo project](https://github.com/ferblaca/OpenRewriteDemo) with the rewrite plugin configured for the above active recipe which contains 6 modules, whose [pom.xml](https://github.com/ferblaca/OpenRewriteDemo/blob/main/pom.xml) of type `pom` has Spring Boot parent version 2.6.5 configured. When running `mvn rewrite:run` the expected result is to update the Spring-boot parent version to `2.6.7`, but this is not the case.

Only when I run the command with -N (`mvn rewrite:run -N`) or comment out all or most of the [modules](https://github.com/ferblaca/OpenRewriteDemo/blob/main/pom.xml#L21-L26), the recipe correctly updates the Spring Boot parent version to `2.6.7`.

Can you check it? thank you very much.

**Comment by tkvangorder** [comment]
Thank you for the detailed example, we will look into this.

**Comment by dmt** [comment]
Just a note if this is relevant… I've had some similar issues that got fixed with version 4.24.0 of the maven plugin

**Comment by ferblaca** [comment]
> Just a note if this is relevant… I've had some similar issues that got fixed with version 4.24.0 of the maven plugin

thanks @dmt ! I'll try it and see if it has been corrected ;)

**Comment by ferblaca** [comment]
@tkvangorder, 

as @dmt says, when doing the same test now with version `4.24.0` of the rewrite plugin it works correctly!!

Can you close the issue thank you!

**Comment by JoseLora** [comment]
@tkvangorder This is happening with last version (rewrite-maven-plugin:4.44.0). I attach sample minimal multi-module project.
TestRecipe code:

```java
package com.test;

import org.openrewrite.Recipe;
import org.openrewrite.maven.ChangeParentPom;

public class TestRecipe extends Recipe {

  public TestRecipe() {
    this.doNext(new ChangeParentPom("org.springframework.boot", null, "spring-boot-starter-parent", null,
        "2.6.7", null, true, null));
  }

  @Override
  public String getDisplayName() {
    return "Migration test";
  }

  @Override
  public String getDescription() {
    return "Migration test.";
  }

}
```
When launching the recipe from a custom recipe library jar via:

```
mvn -U  -org.openrewrite.maven:rewrite-maven-plugin:run -D"rewrite.recipeArtifactCoordinates=com.xxxxxx.amgorwr:amigaorwrecipes:0.1.0-SNAPSHOT" -D"rewrite.activeRecipes=com.test.TestRecipe"
```
It does nothing!

But, if I add maven -N option, then it works flawlessly. It seems a problem with multi-module projects.

Maven version: 3.8.4
Java version: 17
[openrewrite-test.zip](https://github.com/openrewrite/rewrite/files/11295099/openrewrite-test.zip)


Could you please check this?

**Comment by tylervangorder-8451** [comment]
I am no longer on the OpenRewrite team, I am going to ping @timtebeek on this issue. Thanks for reporting!

**Comment by timtebeek** [comment]
Hi all; we just pushed a new [snapshot version of the Maven plugin](https://docs.openrewrite.org/reference/snapshot-instructions#maven-instructions) with a [potential fix](https://github.com/openrewrite/rewrite-maven-plugin/pull/552). Would be so kind as to give it another try with version `4.46.0-SNAPSHOT`?

```xml
<build>
    <plugins>
        <plugin>
            <groupId>org.openrewrite.maven</groupId>
            <artifactId>rewrite-maven-plugin</artifactId>
            <version>4.46.0-SNAPSHOT</version>
        </plugin>
    </plugins>
</build>
```
Given that it's been non deterministic, it's been rather hard to pin down. We did see this being more prevalent with a larger number of submodules, so we're hoping it was the linked ordering issue. Would you mind running a few times even if it works the first time?

**Comment by nmck257** [comment]
FYI @timtebeek -- I can help test out that SNAPSHOT fix first thing tomorrow.
I was troubleshooting this same issue today, narrowed it down to a `rewrite-maven-plugin` thing (saw a module with itself as its parent in the resolution result), and just saw your commits.

**Comment by nmck257** [comment]
Fix looks good, after running 3 times across 2 repos. The "Project [xyz] parsing source files..." messages appear in a consistent, sensible order (whereas previously the order was chaotic), and the ChangeParentPom recipe works as expected.

Would love a patch release of the maven plugin with this fix (perhaps after positive feedback on #3219)

**Comment by timtebeek** [comment]
Great to hear @nmck257 , thanks for confirming so quickly. I'm getting mixed reports in #3219, but there might a different underlying issue there. I'll see about getting a patch release out, with help if needed. Would be great to close this one finally.

**Comment by timtebeek** [comment]
We've just released [v4.45.1 of the plugin](https://github.com/openrewrite/rewrite-maven-plugin/releases/tag/v4.45.1), which based on the above resolves this issue. Thanks all for reporting this here, and your patience while we troubleshooted this one.
